aidb/engine/approx_select_engine.py

```diff
@@ -0,0 +1,211 @@
+    for idx, (index, row) in enumerate(score_for_all_df.iterrows()):
+      min_score = 1
+      for or_connected in score_connected:
+        max_score = 0
+        for score_name in or_connected:
+          max_score = max(max_score, row[score_name])
+        min_score = min(min_score, max_score)
+      proxy_score_all_blobs[idx] = min_score
+    return pd.Series(proxy_score_all_blobs, index=score_for_all_df.index)
+
+
+  async def execute_approx_select_query(self, query: Query):
+    query.is_valid_approx_select_query()
+    # generate proxy score for each blob
+    score_for_all_df, score_connected = await self.get_proxy_scores_for_all_blobs(query)
+
+    proxy_score_for_all_blobs = self.score_fn(score_for_all_df, score_connected)
+
+    dataset = self.get_sampled_proxy_blob(proxy_score_for_all_blobs)
+
+    # This is used for parallel test
+    seed = (mp.current_process().pid * np.random.randint(100000, size=1)[0]) % (2**32 - 1)
+    sampled_df = dataset.sample(BUDGET, replace=True, weights=WEIGHT, random_state=seed)
+
+
+    async with self._sql_engine.begin() as conn:
+      satisfied_sampled_results, all_sampled_results = await self.get_inference_results(
+          query,
+          list(sampled_df.index),
+          conn
+      )
+
+      satisfied_sampled_index = list(set(satisfied_sampled_results.index).intersection(set(sampled_df.index)))
+      satisfied_sampled_results = satisfied_sampled_results.loc[satisfied_sampled_index]
+      satisfied_sampled_results = satisfied_sampled_results.join(sampled_df, how='inner')
+      sorted_satisfied_sampled_results = satisfied_sampled_results.sort_values(by=PROXY_SCORE, ascending=False)
+
+      all_sampled_index = list(set(all_sampled_results.index).intersection(set(sampled_df.index)))
+
+      no_result_length = BUDGET - len(sampled_df.loc[all_sampled_index])
+      all_sampled_results = all_sampled_results.loc[all_sampled_index]
+      all_sampled_results = all_sampled_results.join(sampled_df, how='inner')
+
+      total_length = no_result_length + len(all_sampled_results)
+
+      recall_target = query.recall_target
+      confidence = query.confidence / 100
+      tau_modified = self.tau_modified(
+          sorted_satisfied_sampled_results,
+          recall_target,
+          confidence,
+          total_length
+      )
+
+      R1 = sorted_satisfied_sampled_results.index
+      R2 = dataset[dataset[PROXY_SCORE] >= tau_modified].index
+
+      additional_samples = list(set(R1).union(set(R2)))
+
+      print('num_samples', len(additional_samples))
+      logging.info(f'num_samples: {len(additional_samples)}')
+
+      additional_satisfied_sampled_results,  additional_all_sampled_results = await self.get_inference_results(
+          query,
+          additional_samples,
+          conn
+      )
+
+    return additional_satisfied_sampled_results
+
+
+  async def get_inference_results(self, query:Query, sampled_index: List[int], conn):
```

Comment:
What does this do? Can you call the full scan engine?

Reply:
This function run inference on sampled blob and retrieve results. Currently, we can't directly call the full scan engine because IN operator is not supported in WHERE clause. I will add a TODO and fix it after IN operator is supported.